src/rp_ml_workflow_logger/logger.py

```diff
@@ -0,0 +1,128 @@
+import os
+import json
+import threading
+import pandas as pd
+import networkx as nx
+from datetime import datetime
+from pymongo import MongoClient
+
+class MLWorkFlowLogger:
+    _instance = None
+
+    def __new__(cls, *args, **kwargs):
+        if cls._instance is None:
+            cls._instance = super(MLWorkFlowLogger, cls).__new__(cls)
+            cls._instance.__init__(*args, **kwargs)
+        return cls._instance
+
+    def __init__(self, log_dir='logs', graph=None, db_config=None):
+        self.log_dir = log_dir                                       # Initialize Logs
+        os.makedirs(self.log_dir, exist_ok=True)
+        self.current_run_dir = None
+        self.graph = graph if graph is not None else nx.DiGraph()    # Initialize Networkx graph
+        self.db_config = db_config                                   # Initialize Database
+        self.local_mode = db_config is None
+        self.client = None
+        self.db = None
+        if not self.local_mode:
+            self._setup_database()
+
+
+    def _setup_database(self):
+        if self.db_config['type'] == 'mongodb':
```

Comment:
db_config should a dataclass that has all the fixed parameters. There should be no "guessing" of whats in the dictionary for configs.